File: problems/树/144.二叉树的前序遍历.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

vector<int> preorderTraversal(TreeNode* root) {
    vector<int> result;
    stack<TreeNode*> st;
    if (root) st.push(root);
    while (!st.empty()) {
        TreeNode* node = st.top(); st.pop();
        result.push_back(node->val);
        if (node->right) st.push(node->right);
        if (node->left)  st.push(node->left);
    }
    return result;
}
int main() {
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(3);
    for (int v : preorderTraversal(root)) cout << v << " "; cout << endl;  // 1 2 3
    for (int v : preorderTraversal(new TreeNode(1))) cout << v << " "; cout << endl;  // 1
    for (int v : preorderTraversal(nullptr)) cout << v << " "; cout << endl;  // (empty)
    return 0;
}

File: problems/树/144.二叉树的前序遍历.py
#!/usr/bin/env python3
"""
144. 二叉树的前序遍历
"""
from typing import List, Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def preorder_traversal(root: Optional[TreeNode]) -> List[int]:
    result = []
    stack = [root]
    while stack:
        node = stack.pop()
        if node:
            result.append(node.val)
            stack.append(node.right)
            stack.append(node.left)
    return result

if __name__ == "__main__":
    root1 = TreeNode(1, None, TreeNode(2, TreeNode(3), None))
    print(preorder_traversal(root1))  # [1, 2, 3]
    print(preorder_traversal(TreeNode(1)))  # [1]
    print(preorder_traversal(None))  # []
